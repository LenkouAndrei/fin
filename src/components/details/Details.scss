@import "../../styles/variables.scss";

.details {
    display: flex;
    flex-flow: column;
    position: absolute;
    width: 100%;

    &-wrapper {
        position: relative;
    }

    &__list {
        padding: 15px 0;
        display: flex;
        flex-flow: column;
        align-items: center;
        background-color: lighten($d-green, 30);

        &-item {
            width: 800px;
            font-size: 1.6rem;
            color: $white;
            font-weight: 500;
        }
    }

    &__separator {
        height: 3px;
        background: $d-hakki;
    }

    &__toggler {
        align-self: center;
    }
}