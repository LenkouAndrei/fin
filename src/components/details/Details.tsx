import React, { useState } from 'react';
import { Button, BtnSize, BtnShape } from '../button/Button';
import { ISpending } from '../../services/spendings.service';
import './Details.scss';

const emptyList = (
  <li className="details__list-item--empty">Storage is empty</li>
);

interface IDetails {
  spendings: ISpending[];
}

export const Details: React.FC<IDetails> = ({ spendings }: IDetails) => {
  const [isOpened, setIsOpened] = useState(false);
  const toggle = () => {
    setIsOpened(!isOpened);
  };

  const items = spendings.map(({ spendingName, amount, id }: ISpending) => (
    <li key={id} className="details__list-item">
      <span>{spendingName}</span> - <span>{amount}</span>
    </li>
  ));

  return (
    <div className="details-wrapper">
      <div className="details">
        {isOpened && (
          <>
            <ul className="details__list">
              {items.length > 0 ? items : emptyList}
            </ul>
            <div className="details__separator"></div>
          </>
        )}
        <Button
          className="details__toggler"
          size={BtnSize.small}
          shape={BtnShape.halfsquare}
          onClick={toggle}
        >
          Show Details
        </Button>
      </div>
    </div>
  );
};
