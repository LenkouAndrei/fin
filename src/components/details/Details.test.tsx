import { render, screen, waitFor } from '@testing-library/react';
import { Details } from './Details';

const mockSpendings = [
  { id: 'AAA-1', spendingName: 'Test-1', amount: 20 },
  { id: 'AAA-2', spendingName: 'Test-2', amount: 30 },
];

describe('<Details />', () => {
  it('should render Details', () => {
    const { asFragment } = render(<Details spendings={mockSpendings} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain toggle button', async () => {
    render(<Details spendings={[]} />);
    const toggleBtn = screen.getByText('Show Details');
    expect(toggleBtn).toBeTruthy();
  });

  it('should show details on toggleBtn click', async () => {
    const { container } = render(<Details spendings={[]} />);
    const toggleBtn = screen.getByText('Show Details');
    toggleBtn.click();
    await waitFor(() => {
      expect(container.querySelector('.details__list')).toBeTruthy();
    });
  });

  it('should hide details on toggleBtn click', async () => {
    const { container } = render(<Details spendings={[]} />);
    const toggleBtn = screen.getByText('Show Details');
    toggleBtn.click();
    toggleBtn.click();
    await waitFor(() => {
      expect(container.querySelector('.details__list')).toBe(null);
    });
  });

  it('should show empty detail on toggleBtn click if spendings are absent', async () => {
    const { container } = render(<Details spendings={[]} />);
    const toggleBtn = screen.getByText('Show Details');
    toggleBtn.click();
    await waitFor(() => {
      expect(
        container.querySelector('.details__list-item--empty')
      ).toBeTruthy();
    });
  });

  it('should show details on toggleBtn click if spendings are present', async () => {
    const { container } = render(<Details spendings={mockSpendings} />);
    const toggleBtn = screen.getByText('Show Details');
    toggleBtn.click();
    await waitFor(() => {
      expect(container.querySelector('.details__list-item')).toBeTruthy();
    });
  });
});
