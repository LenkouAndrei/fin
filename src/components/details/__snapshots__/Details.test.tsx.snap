// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Details /> should render Details 1`] = `
<DocumentFragment>
  <div
    class="details-wrapper"
  >
    <div
      class="details"
    >
      <button
        class="btn halfsquare  details__toggler small"
      >
        Show Details
      </button>
    </div>
  </div>
</DocumentFragment>
`;
